### ISSUE TITLE
Translation fails if custom module translations are used

### ISSUE THREAD

**Issue by magicsunday** [fisharebest/webtrees#2329]
In my custom module I added the "customTranslations" method

```
public function customTranslations(string $language): array
{
	return (new Translation(
		'modules_v4/module-name/resources/lang/' . $language . '/messages.mo'
	))->asArray();
}

```

to load the translations from an external file. Currently I provide only a german language file. If I now try to switch the language to an not existing one, I get:

```
ErrorException: fopen(modules_v4/module-name/resources/lang/en-US/messages.mo): failed to open stream: No such file or directory in file /vendor/fisharebest/localization/src/Translation.php on line 44 Stack trace: 

1. ErrorException->() /vendor/fisharebest/localization/src/Translation.php:44 
2. fopen() /vendor/fisharebest/localization/src/Translation.php:44 
3. Fisharebest\Localization\Translation->__construct() /modules_v4/module-name/src/Module.php:143 
4. MagicSunday\Webtrees\Module\Module->customTranslations() /app/I18N.php:453 
5. Fisharebest\Webtrees\I18N->init() /app/Http/Controllers/UserController.php:80 
6. Fisharebest\Webtrees\Http\Controllers\UserController->language() [internal]:0 
7. ReflectionMethod->invoke() /app/Application.php:45 
8. Fisharebest\Webtrees\Application->dispatch() /index.php:175 
9. {closure}() /app/Module/HitCountFooterModule.php:185 
10. Fisharebest\Webtrees\Module\HitCountFooterModule->handle() /index.php:161 
11. {closure}() /app/Http/Middleware/CheckCsrf.php:60 
12. Fisharebest\Webtrees\Http\Middleware\CheckCsrf->handle() /index.php:161 
13. {closure}() /app/Http/Middleware/UseTransaction.php:43 
14. Fisharebest\Webtrees\Http\Middleware\UseTransaction->handle() /index.php:161 
15. {closure}() /app/Http/Middleware/DebugBarData.php:41 
16. Fisharebest\Webtrees\Http\Middleware\DebugBarData->handle() /index.php:161 
17. {closure}() /app/Http/Middleware/BootModules.php:59 
18. Fisharebest\Webtrees\Http\Middleware\BootModules->handle() /index.php:161 
19. {closure}() /app/Http/Middleware/UseTheme.php:71 
20. Fisharebest\Webtrees\Http\Middleware\UseTheme->handle() /index.php:161 
21. {closure}() /app/Http/Middleware/UseLocale.php:52 
22. Fisharebest\Webtrees\Http\Middleware\UseLocale->handle() /index.php:161 
23. {closure}() /app/Http/Middleware/UseTree.php:58 
24. Fisharebest\Webtrees\Http\Middleware\UseTree->handle() /index.php:161 
25. {closure}() /app/Http/Middleware/UseSession.php:59 
26. Fisharebest\Webtrees\Http\Middleware\UseSession->handle() /index.php:161 
27. {closure}() /app/Http/Middleware/UseFilesystem.php:50 
28. Fisharebest\Webtrees\Http\Middleware\UseFilesystem->handle() /index.php:161 
29. {closure}() /app/Http/Middleware/CheckForMaintenanceMode.php:48 
30. Fisharebest\Webtrees\Http\Middleware\CheckForMaintenanceMode->handle() /index.php:161 
31. {closure}() /index.php:178
```

If a second custom module is installed, the switch will fail even if the first one provides the desired language.

**Comment by magicsunday** [comment]
You can test it with https://github.com/magicsunday/webtrees-fan-chart or/and https://github.com/magicsunday/webtrees-pedigree-chart

**Comment by fisharebest** [comment]
Try creating `resourcesFolder()` as follows

```
    /**
     * Where does this module store its resources
     *
     * @return string
     */
    public function resourcesFolder(): string
    {
        return __DIR__ . '/resources/';
    }
```

Then use

`$this->resourcesFolder() . 'lang/' . $language . '/messages.mo'`

I have some local changes that use this `resourcesFolder()` function for other things, such as asset URLs, etc.

**Comment by magicsunday** [comment]
Makes no difference.

**Comment by fisharebest** [comment]
Your function `customTranslations(string $language): array` (https://github.com/magicsunday/webtrees-fan-chart/blob/master/src/FanChartModule.php#L150) is trying to use a non-existing file.

Compare with the example (https://github.com/fisharebest/webtrees/blob/master/modules_v4/example.disable/module.php#L111) which knows which files are available.

You could use  `file_exists()` instead of the `switch`.  I used `switch` for performance.

**Comment by magicsunday** [comment]
Ok. Checking for the file does the trick. In my opinion implementing the translation stuff seems rather complicated. Maybe webtrees should check for itself if translations in a given module directory (e.g. resources/lang/*) and automatically include the files, so you does not need to add a method to each module?

**Comment by ric2016** [comment]
Loading translation from custom modules currently doesn't seem to work at all (out of the box), because in all calls to `I18N::init()`, `$setup` is directly or implicitly true, which prevents translations from custom modules getting loaded.

At least the init in `UseLocale.php` is probably supposed to be

`define('WT_LOCALE', I18N::init('', $tree, false));`

now? Other calls may also have to be adjusted.

**Comment by fisharebest** [comment]
I think this was fixed in 6b96fc9048544acaaa3a51f75cb44c5322b5030d

> Maybe webtrees should check for itself if translations in a given module directory (e.g. resources/lang/*)

The would involve lots of calls to `file_exists()`.  I know this is cached by PHP, but it is much slower when checking for non-existant files.

**Comment by ric2016** [comment]
> I think this was fixed in [6b96fc9](https://github.com/fisharebest/webtrees/commit/6b96fc9048544acaaa3a51f75cb44c5322b5030d)

Yes - I just wasn't up-to-date, sorry.
